### PR TITLE
Enable offlineAnalysis return list of anomalies

### PR DESCRIPTION
**Pull request by cyy0714**
1. Rename offline_analysis to offlineAnalysis

2. Enable offlineAnalysis to return list of anomalies so that front-end can take the offline analysis result.

3. Modify components in the detectionJobScheduler for code reuse.

4. Autotune configuration application will skip if configuration is empty.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java [hunk 1]
```diff
@@ -438,37 +437,33 @@ public void synchronousBackFill(long functionId, DateTime backfillStartTime, Dat
...
-              jobDTO.setStatus(JobConstants.JobStatus.FAILED);
-              jobDAO.save(jobDTO);
-            }
-
-            return;
-          }
-          hasFailedTask = true;
-          LOG.warn("The backfill of anomaly function {} (task id: {}) failed. Retry count: {}", functionId,
-              jobExecutionId, retryCounter);
+          failedTaskIds.add(taskDTO.getId());
```
Comment (cecilynie) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java]: If a task always failed, will this cause an infinite loop. Should think about ways to give a hard limit for failed jobs to restart

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java [hunk 2]
```diff
@@ -495,9 +495,11 @@ public Response applyAutotuneConfig(@PathParam("autotuneConfigId") @NotNull long
     // if auto tune method is ALERT_FILTER_LOGISITC_AUTO_TUNE or INITIATE_ALERT_FILTER_LOGISTIC_AUTO_TUNE, alert filter is to be updated
     if (autotuneConfigDTO.getAutotuneMethod() != EXHAUSTIVE) {
       targetFunction.setAlertFilter(autotuneConfigDTO.getConfiguration());
-    } else{
+    } else {
       // Update function configuration
-      targetFunction.updateProperties(autotuneConfigDTO.getConfiguration());
+      if (autotuneConfigDTO.getConfiguration()!= null && !autotuneConfigDTO.getConfiguration().isEmpty()) {
```
Comment (cecilynie) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java]: if new autotuneConfigDTO is empty, are we designed to not update target Function? Let's log this difference, the original function setting is already optimal after tuning.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java [hunk 3]
```diff
@@ -438,37 +437,33 @@ public void synchronousBackFill(long functionId, DateTime backfillStartTime, Dat
...
+  public JobStatus waitForJobDone(long jobExecutionId) {
     TaskManager taskDAO = DAO_REGISTRY.getTaskDAO();
     JobManager jobDAO = DAO_REGISTRY.getJobDAO();
-    JobDTO jobDTO = null;
+    JobDTO jobDTO;
     List<TaskDTO> scheduledTaskDTO = taskDAO.findByJobIdStatusNotIn(jobExecutionId, TaskConstants.TaskStatus.COMPLETED);
+    long functionId = jobDAO.findById(jobExecutionId).getAnomalyFunctionId();
 
-    int retryCounter = 0;
     while (scheduledTaskDTO.size() > 0) {
```
Comment (cecilynie) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java]: change while to if!

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java [hunk 4]
```diff
@@ -287,9 +288,10 @@ public Response generateAnomaliesInRange(@PathParam("id") @NotNull String id,
   }
 
   @POST
-  @Path("/{id}/offline_analysis")
+  @Path("/{id}/offlineAnalysis")
   public Response generateAnomaliesInTrainingData(@PathParam("id") @NotNull long id,
       @QueryParam("time") String analysisTimeIso) throws Exception {
+    long currentTime = System.currentTimeMillis();
```
Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: Good catch. Removed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java (hunk 1 above)
Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java]: Solved. The scheduledTaskDTO is updated at the end of the loop.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java (hunk 3 above)
Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java]: Fixed.

**Pull request merged**